### PR TITLE
#164139695 Add search and filter functionality

### PR DESCRIPTION
**Pull request by codinger41**
#### What does this PR do?
Adds the feature for users to be able to search and filter articles.
#### Description of Task to be completed?
Users should be able to search for articles. While searching, users should be able to filter their search results. 
This new endpoint takes these queries:
- `term` the search keyword
- `categoryId` the category the user wants to search from. We'll provide this based on what the user selects on the front end.
- `author` the username or name of the author they want to search from.
- `startDate` and `endDate` the start and end dates for a date range filter
- `tags` a string of tags they want to search from.
#### How should this be manually tested?
- run  `npm run db-migrate` to get all database migration updates.
- make a GET request to the `/search` route with the above query parameters.
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/164139695

#### Questions:
N/A

### CODE REVIEW

File: controllers/articles.js [hunk 1]
```diff
@@ -34,4 +40,119 @@ export default class ArticleController {
...
+          attributes: ['username', 'email', 'name', 'bio'],
+          as: 'author'
+        }]
+      });
+      return res.status(200).json({
+        results,
+        success: true
+      });
+    } catch (error) {
+      return res.status(200).json({
```
Comment (olajide1234) [inline controllers/articles.js]: Please consider returning a 404 here.

File: controllers/articles.js [hunk 2]
```diff
@@ -34,4 +40,118 @@ export default class ArticleController {
...
+        }]
+      });
+      return res.status(200).json({
+        success: true,
+        article: article.dataValues
+      });
+    } catch (error) {
+      return res.status(404).json({
+        success: false,
+        errors: ['Article not found.'],
```
Comment (Dubby20) [inline controllers/articles.js]: This should be an error, not errors

Comment (leksyib) [reply controllers/articles.js]: Our convention uses `errors` and it's an array. If I change it to `error`, we won't be consistent and when we're working on the front end, we won't know the exact one to use. Let's keep being consistent so it'll be easy for us. Thanks.

**Review by Dubby20 — APPROVED** [review]
Looks good to merge

**Review by kleva-j — APPROVED** [review]

**Review by olajide1234 — APPROVED** [review]

**Pull request closed without merging**